common/src/api/external/mod.rs

```diff
@@ -2126,6 +2148,376 @@ impl std::fmt::Display for Digest {
     }
 }
 
+/// An address lot and associated blocks resulting from creating an address lot.
+#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
+pub struct AddressLotCreateResponse {
+    /// The address lot that was created.
+    pub lot: AddressLot,
+
+    /// The address lot blocks that were created.
+    pub blocks: Vec<AddressLotBlock>,
+}
+
+/// Represents an address lot object, containing the id of the lot that can be
+/// used in other API calls.
+#[derive(
+    ObjectIdentity, Clone, Debug, Deserialize, JsonSchema, Serialize, PartialEq,
+)]
+pub struct AddressLot {
+    #[serde(flatten)]
+    pub identity: IdentityMetadata,
+}
+
+/// An address lot block is a part of an address lot and contains a range of
+/// addresses. The range is inclusive.
+#[derive(Clone, Debug, Deserialize, JsonSchema, Serialize, PartialEq)]
+pub struct AddressLotBlock {
+    /// The id of the address lot block.
+    pub id: Uuid,
+
+    /// The first address of the block (inclusive).
+    pub first_address: IpAddr,
+
+    /// The last address of the block (inclusive).
+    pub last_address: IpAddr,
+}
+
+/// A loopback address is an address that is assigned to a rack switch but is
+/// not associated with any particular port.
+#[derive(Clone, Debug, Deserialize, JsonSchema, Serialize, PartialEq)]
+pub struct LoopbackAddress {
+    /// The id of the loopback address.
+    pub id: Uuid,
+
+    /// The address lot block this address came from.
+    pub address_lot_block_id: Uuid,
+
+    /// The id of the rack where this loopback address is assigned.
+    pub rack_id: Uuid,
+
+    /// Switch location where this loopback address is assigned.
+    pub switch_location: String,
+
+    /// The loopback IP address and prefix length.
+    pub address: IpNet,
+}
+
+/// A switch port represents a physical external port on a rack switch.
+#[derive(Clone, Debug, Deserialize, JsonSchema, Serialize, PartialEq)]
+pub struct SwitchPort {
+    /// The id of the switch port.
+    pub id: Uuid,
+
+    /// The rack this switch port belongs to.
+    pub rack_id: Uuid,
```

Comment:
Something to note (that I have anxiety about) is that we currently don't have a deterministic way to associate racks and ids. I mean, a rack is technically just a piece of metal with removal hardware in it. There's a rack of theseus problem here by which if enough of the hardware is changed and moved over to different places, who says what rack it's in anymore? It makes me wonder if the more deterministic thing isn't to just have a switch UUID which its associated to.

Reply:
https://github.com/oxidecomputer/omicron/issues/3066